samples/bluetooth/bap_broadcast_sink/src/audio_codec.c

```diff
@@ -0,0 +1,172 @@
+ * 480 bytes correspond to 240 PCM samples per block. This value was
+ * chosen as a compromise between:
+ *  - keeping latency reasonably low (smaller blocks)
+ *  - avoiding excessive interrupt / DMA overhead (larger blocks).
+ * If the PCM format (sample width or number of channels) is changed,
+ * this constant should be revisited so that the block size remains
+ * appropriate for the target use case and hardware.
+ */
+#define CODEC_BLOCK_SIZE 480
+
+/*
+ * Default speaker/output volume sent via AUDIO_PROPERTY_OUTPUT_VOLUME.
+ *
+ * The valid numeric range and exact mapping to dB are codec‑dependent;
+ * refer to the specific codec driver / datasheet for details. The value
+ * 15 was selected as a conservative, comfortable default for this sample
+ * application. Adjust as needed to match the desired default output level.
+ */
+#define SPEAKER_VOL      15
+
+/*
+ * Total size of the ring buffer used to queue audio data for the codec.
+ *
+ * The buffer is sized to hold 20 CODEC_BLOCK_SIZE blocks. Having multiple
+ * blocks queued reduces the likelihood of underruns when the producer is
+ * briefly delayed, while still keeping overall latency and RAM usage
+ * bounded. Increasing the multiplier adds buffering (and latency / RAM
+ * consumption); decreasing it reduces buffering but makes underruns more
+ * likely on a busy system.
+ */
+#define RING_BUF_SIZE    (CODEC_BLOCK_SIZE * 20)
+
+#if DT_NODE_HAS_STATUS_OKAY(DT_ALIAS(codec0))
+static uint8_t ring_buffer[RING_BUF_SIZE];
+static struct ring_buf rb;
+static bool codec_configed;
+static uint8_t block_data[CODEC_BLOCK_SIZE];
+static const struct device *g_dev;
+
+static void tx_done(const struct device *dev, void *user_data)
+{
+	uint32_t avail;
+	uint32_t read;
+
+	avail = ring_buf_size_get(&rb);
+	if (avail < CODEC_BLOCK_SIZE) {
+		LOG_WRN("Buffer empty\n");
+		return;
+	}
+	read = ring_buf_get(&rb, block_data, CODEC_BLOCK_SIZE);
+	if (read != CODEC_BLOCK_SIZE) {
+		LOG_WRN("Failed to read full block from ring buffer: %u bytes\n", read);
+		return;
+	}
+	audio_codec_write(dev, block_data, CODEC_BLOCK_SIZE);
+}
+int audio_codec_open(void)
+{
+	ring_buf_init(&rb, RING_BUF_SIZE, ring_buffer);
+	g_dev = DEVICE_DT_GET(DT_ALIAS(codec0));
+	if (!device_is_ready(g_dev)) {
+		LOG_ERR("Could not open codec device\n");
+		return -EIO;
+	}
+	return 0;
+}
+int audio_codec_cfg(uint32_t samplerate)
+{
+	int ret;
+	audio_property_value_t val;
+	struct audio_codec_cfg cfg = {
+		.dai_type = AUDIO_DAI_TYPE_PCM,
+		.dai_cfg.pcm.dir = AUDIO_DAI_DIR_TX,
+		.dai_cfg.pcm.pcm_width = AUDIO_PCM_WIDTH_16_BITS,
+		.dai_cfg.pcm.channels = 1,
+		.dai_cfg.pcm.block_size = CODEC_BLOCK_SIZE,
+		.dai_cfg.pcm.samplerate = samplerate,
+	};
+
+	if (codec_configed) {
+		LOG_WRN("Already configured\n");
+		return -EALREADY;
+	}
+	LOG_INF("codec: samplerate=%d block_size=%d\n", samplerate, CODEC_BLOCK_SIZE);
+	audio_codec_register_done_callback(g_dev, tx_done, NULL, NULL, NULL);
```

Comment:
Where is `audio_codec_register_done_callback` defined? Can't find it anywhere :s